<a href="https://colab.research.google.com/github/SamShinwari/Python-Programming/blob/main/HousePricePrediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **📘 Hands-on Exercise 2: Data Preparation and Preprocessing (Google Colab)**

In this exercise, I will perform data preparation and preprocessing using Python and the Pandas library in Google Colab on the dataset HousePricePrediction.csv.

I will start by loading the dataset into a Pandas DataFrame and exploring its structure to understand the data better. This will include displaying sample rows, checking the dataset shape, identifying data types, and generating summary statistics.

After exploration, I will clean and preprocess the dataset by:



*   Handling missing values appropriately
*   Removing duplicate records (if any)
*   Selecting relevant features for analysis
*   Converting categorical variables into numerical format

The main objective of this exercise is to prepare a clean, consistent, and machine-learning-ready dataset for further analysis and model building.

In [ ]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
import numpy as np
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


#Load the Dataset

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/HousePricePrediction.csv')

### Display first 10 rows

In [ ]:
print("First 10 Rows of Dataset:\n")
print(df.head(10))

First 10 Rows of Dataset:

   Id  MSSubClass MSZoning  LotArea LotConfig BldgType  OverallCond  \
0   0          60       RL     8450    Inside     1Fam            5   
1   1          20       RL     9600       FR2     1Fam            8   
2   2          60       RL    11250    Inside     1Fam            5   
3   3          70       RL     9550    Corner     1Fam            5   
4   4          60       RL    14260       FR2     1Fam            5   
5   5          50       RL    14115    Inside     1Fam            5   
6   6          20       RL    10084    Inside     1Fam            5   
7   7          60       RL    10382    Corner     1Fam            6   
8   8          50       RM     6120    Inside     1Fam            5   
9   9         190       RL     7420    Corner   2fmCon            6   

   YearBuilt  YearRemodAdd Exterior1st  BsmtFinSF2  TotalBsmtSF  SalePrice  
0       2003          2003     VinylSd         0.0        856.0   208500.0  
1       1976          1976     MetalS

# =================================
# 3. Data Exploration
# =================================


#### Shape of dataset

In [ ]:
print("\nShape of Dataset:")
print(df.shape)


Shape of Dataset:
(2919, 13)


#### Data types of columns

In [ ]:
print("\nData Types:\n")
print(df.dtypes)


Data Types:

Id                int64
MSSubClass        int64
MSZoning         object
LotArea           int64
LotConfig        object
BldgType         object
OverallCond       int64
YearBuilt         int64
YearRemodAdd      int64
Exterior1st      object
BsmtFinSF2      float64
TotalBsmtSF     float64
SalePrice       float64
dtype: object


#### Summary statistics

In [ ]:
print("\nSummary Statistics:\n")
print(df.describe())


Summary Statistics:

                Id   MSSubClass        LotArea  OverallCond    YearBuilt  \
count  2919.000000  2919.000000    2919.000000  2919.000000  2919.000000   
mean   1459.000000    57.137718   10168.114080     5.564577  1971.312778   
std     842.787043    42.517628    7886.996359     1.113131    30.291442   
min       0.000000    20.000000    1300.000000     1.000000  1872.000000   
25%     729.500000    20.000000    7478.000000     5.000000  1953.500000   
50%    1459.000000    50.000000    9453.000000     5.000000  1973.000000   
75%    2188.500000    70.000000   11570.000000     6.000000  2001.000000   
max    2918.000000   190.000000  215245.000000     9.000000  2010.000000   

       YearRemodAdd   BsmtFinSF2  TotalBsmtSF      SalePrice  
count   2919.000000  2918.000000  2918.000000    1460.000000  
mean    1984.264474    49.582248  1051.777587  180921.195890  
std       20.894344   169.205611   440.766258   79442.502883  
min     1950.000000     0.000000     0.00

# =================================
# 4. Data Cleaning
# =================================



### Check missing values

In [ ]:
print("\nMissing Values:\n")
print(df.isnull().sum())


Missing Values:

Id                 0
MSSubClass         0
MSZoning           4
LotArea            0
LotConfig          0
BldgType           0
OverallCond        0
YearBuilt          0
YearRemodAdd       0
Exterior1st        1
BsmtFinSF2         1
TotalBsmtSF        1
SalePrice       1459
dtype: int64


#### Handle missing values

In [ ]:
# Numerical columns -> fill with mean
num_cols = df.select_dtypes(include=['int64', 'float64']).columns

for col in num_cols:
    df[col].fillna(df[col].mean(), inplace=True)

# Categorical columns -> fill with mode
cat_cols = df.select_dtypes(include=['object']).columns

for col in cat_cols:
    df[col].fillna(df[col].mode()[0], inplace=True)

print("\nMissing values handled successfully.")

# Check duplicate records
print("\nDuplicate Records:", df.duplicated().sum())

# Remove duplicates if any
df.drop_duplicates(inplace=True)

print("Duplicates removed successfully.")





Missing values handled successfully.

Duplicate Records: 0
Duplicates removed successfully.


/tmp/ipykernel_8263/1008593984.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(df[col].mean(), inplace=True)
/tmp/ipykernel_8263/1008593984.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try usi

# =================================
# 5. Feature Selection
# =================================

In [ ]:
# Remove unnecessary column 'Id'
if 'Id' in df.columns:
    df.drop('Id', axis=1, inplace=True)

print("\nColumns after Feature Selection:\n")
print(df.columns)


Columns after Feature Selection:

Index(['MSSubClass', 'MSZoning', 'LotArea', 'LotConfig', 'BldgType',
       'OverallCond', 'YearBuilt', 'YearRemodAdd', 'Exterior1st', 'BsmtFinSF2',
       'TotalBsmtSF', 'SalePrice'],
      dtype='object')


# =================================
# 6. Data Preprocessing
# =================================

In [ ]:

# Convert categorical variables into numerical form
label_encoder = LabelEncoder()

for col in cat_cols:
    if col in df.columns:
        df[col] = label_encoder.fit_transform(df[col])

# Display processed dataset preview
print("\nPreprocessed Dataset Preview:\n")
print(df.head())

# Final dataset info
print("\nFinal Dataset Shape:")
print(df.shape)


Preprocessed Dataset Preview:

   MSSubClass  MSZoning  LotArea  LotConfig  BldgType  OverallCond  YearBuilt  \
0          60         3     8450          4         0            5       2003   
1          20         3     9600          2         0            8       1976   
2          60         3    11250          4         0            5       2001   
3          70         3     9550          0         0            5       1915   
4          60         3    14260          2         0            5       2000   

   YearRemodAdd  Exterior1st  BsmtFinSF2  TotalBsmtSF  SalePrice  
0          2003           12         0.0        856.0   208500.0  
1          1976            8         0.0       1262.0   181500.0  
2          2002           12         0.0        920.0   223500.0  
3          1970           13         0.0        756.0   140000.0  
4          2000           12         0.0       1145.0   250000.0  

Final Dataset Shape:
(2919, 12)


# **📘 Hands-on Exercise 3: Machine Learning Model Building and Evaluation**

###In this exercise, I will build and evaluate a Machine Learning model using the cleaned dataset prepared in Hands-on Exercise 2.

I will begin by splitting the dataset into training and testing sets. Then, I will train a Linear Regression model to predict house prices based on the available features.

After training the model, I will use it to make predictions on the test data. I will evaluate the model using the following performance metrics:



*   Mean Absolute Error (MAE)
*   Mean Squared Error (MSE)
*   Root Mean Squared Error (RMSE)
*   R² Score


Finally, I will analyze the results to determine how well the model is performing and whether it needs further improvement.

The objective of this exercise is to help me understand the complete machine learning workflow, from data preparation to model evaluation.

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

#### **Define Features and Target**

In [ ]:
X = df.drop('SalePrice', axis=1)
y = df['SalePrice']

print("Features Shape:", X.shape)
print("Target Shape:", y.shape)

Features Shape: (2919, 11)
Target Shape: (2919,)


#### **Train-Test Split (80% Training, 20% Testing)**

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.20,
    random_state=42
)

print("Training Data Shape:", X_train.shape)
print("Testing Data Shape:", X_test.shape)

Training Data Shape: (2335, 11)
Testing Data Shape: (584, 11)


### **Train Linear Regression Model**

In [ ]:
model = LinearRegression()
model.fit(X_train, y_train)

print("Model trained successfully.")

Model trained successfully.


#### **Predict House Prices**

In [ ]:
# Predict on test data
y_pred = model.predict(X_test)

print("Predicted House Prices:\n")
print(y_pred[:10])

Predicted House Prices:

[169187.61445926 187016.6161716  162163.12491673 193787.55969266
 202121.21306213 223336.29544111 136359.85368826 206496.34249735
 123964.56663963 196395.33856988]


### **Model Evaluation**

In [ ]:
mae = mean_absolute_error(y_test, y_pred)

mse = mean_squared_error(y_test, y_pred)

rmse = np.sqrt(mse)

r2 = r2_score(y_test, y_pred)

# Print results
print("Model Evaluation Results:\n")

print("Mean Absolute Error (MAE):", mae)

print("Mean Squared Error (MSE):", mse)

print("Root Mean Squared Error (RMSE):", rmse)

print("R² Score:", r2)

Model Evaluation Results:

Mean Absolute Error (MAE): 32540.537785054457
Mean Squared Error (MSE): 2372283670.925819
Root Mean Squared Error (RMSE): 48706.09480266119
R² Score: 0.33488184282838285


### **Interpretation**

In [ ]:
print("\nModel Interpretation:\n")

if r2 > 0.80:
    print("The model performance is very good.")
elif r2 > 0.60:
    print("The model performance is acceptable but can be improved.")
else:
    print("The model performance is weak and needs improvement.")


Model Interpretation:

The model performance is weak and needs improvement.


## **Interpretation of Results**



*   The Linear Regression model was successfully trained using the house price dataset.
*   The dataset was divided into training and testing sets using an 80:20 ratio.

*   The Linear Regression model was evaluated using MAE, MSE, RMSE, and R² Score.
*   The Mean Absolute Error (MAE = 32540.54) indicates that, on average, the model’s predictions are off by around 32,540 units of house price, which is relatively high.
*   The Root Mean Squared Error (RMSE = 48706.09) is also high, showing that the model has significant prediction errors, especially for larger values.
*   The R² Score = 0.3349 means that the model is only able to explain about 33% of the variance in house prices, which is considered low performance.

## **⚠️ Conclusion:**

Overall, the model performance is weak. This indicates that Linear Regression is not capturing the relationship between features and house prices effectively.

## **🔧 Possible Improvements:**
*   Apply better feature engineering (create new meaningful features)
*   Handle outliers in the dataset
*   Try feature scaling (Standardization/Normalization)
*   Use more advanced models such as:
Decision Tree Regressor

1.   Decision Tree Regressor
2.   Random Forest Regressor
3.   Gradient Boosting Regressor
